File: zircon/kernel/lib/mistos/zx/vmar.cc
// Copyright 2024 Mist Tecnologia LTDA. All rights reserved.
// Copyright 2016 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "lib/mistos/zx/vmar.h"

#include <trace.h>
#include <zircon/syscalls.h>

namespace zx {

zx_status_t vmar::allocate(uint32_t options, size_t offset, size_t size, vmar* child,
                           uintptr_t* child_addr) const {
  // Allow for aliasing of |child| to the same container as |this|.
  vmar h;
  zx_status_t status =
      zx_vmar_allocate(get(), options, offset, size, h.reset_and_get_address(), child_addr);
  child->reset(h.release());
  return status;
}

}  // namespace zx

File: src/starnix/lib/ebpf_api/src/maps/vmar.rs
// Copyright 2024 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use std::ops::Deref;

/// A RAII container for a vmar. The vmar mapping will be destroyed when the container is dropped.
/// User must ensure not to use the vmar mapping afterwards.
#[derive(Debug)]
pub struct AllocatedVmar {
    vmar: zx::Vmar,
}

impl AllocatedVmar {
    /// Allocate a new subregion in `parent`. Returns an `AllocatedVmar` that keep the region alive
    /// and the base address of the subregion.
    ///
    /// # Safety
    ///
    /// The returned object will deallocate the region when dropped. It is the responsibility of
    /// the caller to ensure the returned memory is only used as long as this object is alive.
    pub unsafe fn allocate(
        parent: &zx::Vmar,
        offset: usize,
        size: usize,
        flags: zx::VmarFlags,
    ) -> Result<(Self, usize), zx::Status> {
        parent.allocate(offset, size, flags).map(|(vmar, base)| (Self { vmar }, base))
    }
}

impl Drop for AllocatedVmar {
    fn drop(&mut self) {
        // SAFETY:
        //
        // This call is safe only if the user of the `AllocatedVmar` ensured the memory is not used
        // anymore when the object is dropped. This is the contract of the `AllocatedVmar::new`
        // method that is itself unsafe.
        unsafe {
            // Ignore errors. Thete is nothing to do at this point.
            let _ = self.vmar.destroy();
        }
    }
}

impl Deref for AllocatedVmar {
    type Target = zx::Vmar;

    fn deref(&self) -> &Self::Target {
        &self.vmar
    }
}
